catalyst-toolbox-lib/src/recovery/tally.rs

```diff
@@ -0,0 +1,648 @@
+    )
+}
+
+fn committee_id_to_address(id: CommitteeIdDef) -> Address {
+    let id = CommitteeId::from(id);
+    let pk = id.public_key();
+    chain_addr::Address(Discrimination::Production, Kind::Account(pk)).into()
+}
+
+fn voteplans_from_block0(block0: &Block) -> HashMap<VotePlanId, VotePlan> {
+    block0
+        .fragments()
+        .filter_map(|fragment| {
+            if let Fragment::VotePlan(tx) = fragment {
+                let voteplan = tx.as_slice().payload().into_payload();
+                Some((voteplan.to_id(), voteplan))
+            } else {
+                None
+            }
+        })
+        .collect()
+}
+
+/// check that the transaction input/outputs/witnesses is valid for the ballot
+/// * Only 1 input (subsequently 1 witness), no output
+pub(super) fn valid_vote_cast(tx: &TransactionSlice<certificate::VoteCast>) -> bool {
+    if tx.inputs().nb_inputs() != 1
+        || tx.witnesses().nb_witnesses() != 1
+        || tx.outputs().nb_outputs() != 0
+    {
+        return false;
+    }
+
+    true
+}
+
+fn verify_original_tx(
+    spending_counter: SpendingCounter,
+    block0_hash: &HeaderId,
+    sign_data_hash: &TransactionSignDataHash,
+    account: &account::Identifier,
+    witness: &account::Witness,
+    range_check: Range<u32>,
+) -> (bool, u32) {
+    let spending_counter: u32 = <u32>::from(spending_counter);
+    for i in range_check {
+        for op in &[u32::checked_add, u32::checked_sub] {
+            if let Some(new_spending_counter) = op(spending_counter, i) {
+                let tidsc = WitnessAccountData::new(
+                    block0_hash,
+                    sign_data_hash,
+                    SpendingCounter::from(new_spending_counter),
+                );
+                if witness.verify(account.as_ref(), &tidsc) == chain_crypto::Verification::Success {
+                    trace!(
+                        "expected: {} found: {}",
+                        spending_counter,
+                        new_spending_counter
+                    );
+                    return (true, new_spending_counter);
+                }
+            }
+        }
+    }
+    (false, 0)
+}
+
+fn increment_ledger_time_up_to(ledger: Ledger, blockdate: BlockDate) -> Ledger {
+    ledger
+        .begin_block(
+            ledger.get_ledger_parameters(),
+            ledger.chain_length().increase(),
+            blockdate,
+        )
+        .unwrap()
+        .finish(&ConsensusEvalContext::Bft)
+}
+
+pub fn deconstruct_account_transaction<P: chain_impl_mockchain::transaction::Payload>(
+    transaction: &TransactionSlice<P>,
+) -> (P, account::Identifier, account::Witness) {
+    let payload = transaction.payload().into_payload();
+    let account = transaction.inputs().iter().next().unwrap().to_enum();
+
+    let identifier = if let InputEnum::AccountInput(account, _) = account {
+        account.to_single_account().unwrap()
+    } else {
+        panic!("cannot handle utxo inputs");
```

Comment:
Unless this kind of thing is screened in `valid_vote_cast`, someone could break our tally tool by submitting a fragment with UTxO inputs or witnesses (that don't have to be valid). We should plug this in a follow-up PR.